Refactor WeylExtraction

mirenradia opened the pull request:
So I've refactored the old `WeylExtraction` class by separating it into a family (sorry @eugenealim for all the children) of inherited classes as follows:

`SurfaceExtraction` <- `SphericalExtraction` <- `WeylExtraction`

Currently the way `SurfaceExtraction` is generalised is by having a member pointer to a `SurfaceGeometry` abstract base class whose childs should be defined appropriately for the relevant surface geometry. I have provided an example in the form of `SphericalGeometry` which is automatically used in the `SphericalExtraction` class.

There are some also some other changes:
- The old `extraction_params_t` struct in SimulationParametersBase.hpp is now
`SphericalExtraction::params_t` and there is some minor member variable
renaming (`extraction_center` is now just `center`).
- The OpenMP parallelisation in `WeylExtraction::integrate_surface` has
been removed in `SurfaceExtraction::integrate`. I'm not sure this routine takes enough time for this to be helpful.
 - Since you have the option of extracting derivatives of variables on
the surfaces, I have added a `Derivative::name` function to provide
a string for writing out depending on what type of derivative it is
- I have added a SurfaceExtractionTest which just copies the BasicAMRInterpolatorTest (which sets the grid variables to a constant value everywhere and then extracts them) but now extracts the values onto some spheres using `SphericalExtraction` and integrates them over the sphere. Finally it compares with the analytic answer.
- I have written an `IntegrationMethod` class to specify what quadrature (e.g. Trapezium rule, Simpson's rule, etc) to use for the integration over the surface.
- The integration method used for extracting modes in `WeylExtraction` has been changed from using the midpoint rule (second order) to Simpson's rule (fourth order) for the theta integral. This requires an odd number of points (to give an even number of subintervals of the integration range) so I have forced `SimulationParametersBase` to increment the value by 1 if an even number is provided. Also, note that, unlike the midpoint rule, the values at the endpoints/poles (theta=0 and theta=pi) are required. However, these points obviously do not contribute to the integral as the area element (r^2 sin(theta)) vanishes there anyway. They are therefore redundant but are still required because of the way the `SurfaceExtraction::integrate` code is written. I don't think this is too much of a problem.
- Probably some other things I have forgotten...

Some questions I have:
- Currently the way I have generalised `SurfaceExtraction` is by having a member pointer to an abstract `SurfaceGeometry` base class though I could switch to just templating over a `SurfaceGeometry` class instead. This sounds like it would be slightly more performant but I'm not sure it's necessary. Does anyone have any opinions?
- Obviously I had integrating over a sphere in my head when writing `SurfaceExtraction` but I tried to write it in a geometry independent way. Can you find any places where I might have implicitly assumed a sphere which would make it fail to generalise properly to other 2-parameter surfaces?
- The BasicAMRInterpolatorTest is now possibly a bit redundant since the SurfaceExtractionTest tests the `AMRInterpolator` as well. Do you think we should remove it?
- Is it bad that `SimulationParametersBase` now modifies `num_points_theta` if the user sets it to an even number? Is there a better way around this problem?

amelialdrew commented:
Hi, thanks for this. I have written a CylinderExtraction and CylinderGeometry routine which I can add in, what would be the best way to do this with the current CTC/public branches? I also had a couple of suggestions which I would be interested to hear opinions on:

- I would suggest renaming ‘SurfaceExtractionTest’ to ‘SphericalExtractionTest’ as it is currently not a general surface. I wrote a ‘CylinderExtractionTest’ which I could add if necessary. 

- In answer to Miren’s 2nd bullet point on the pull request, from my cylinder tests SurfaceExtraction seems to be geometry independent, which is great. However, there are a few places in the GR implementation where it is specifically coded in for spheres, e.g. SimulationParametersBase defines extraction_params as a SphericalExtraction. Miren has explained that this was intentional, as spheres are the main extraction surface in many GR scenarios and more specifically that my use of a cylinder (for now) does not involve the GR parts. Do you think it would be useful to make it more general at this stage, or to wait until further down the line? Or that it should be left to the user to e.g. edit SimulationParametersBase if they want a different surface?

mirenradia commented:
- I have pushed my latest commit to this branch so if you want to add 1 commit to add your cylinder classes, I'm happy for you to do so. For consistency with `SphericalExtraction` and `SphericalGeometry`, I'd prefer they be called `CylindricalExtraction` and `CylindricalGeometry` respectively rather than `CylinderExtraction` and `CylinderGeometry`.
- I'm happy to rename the current `SurfaceExtractionTest` to `SphericalExtractionTest`. I'm not sure it's necessary or useful to have another extraction test as I mainly wrote it to test the `SurfaceExtraction` class.
- I think the confusion arises because the `SimulationParametersBase` class is misleadingly named as it mostly contains CCZ4/GR parameters. It should be named something more appropriate but otherwise stay as is (i.e. with `SphericalExtraction::params_t extraction_params`).

TaigoFr commented:
Can you change the default center to 'center' we already have and not the {L/2}? I can do this later when the symmetry is added, but it's a redundancy I think

TaigoFr commented on Source/GRChomboCore/SimulationParametersBase.hpp:
```diff
@@ -89,8 +76,14 @@ class SimulationParametersBase : public ChomboParameters
+            pout() << "Parameter: num_points_theta incompatible with Simpson's "
+                   << "rule so increased by 1.\n";
+        }
+        pp.load("extraction_center", extraction_params.center,
```
Can you change the default center to 'center' we already have and not the {L/2}? I can do this later when the symmetry is added, but it's a redundancy I think

mirenradia replied:
Done

TaigoFr commented on Tests/CylindricalExtractionTest/SimulationParameters.hpp:
```diff
@@ -0,0 +1,58 @@
+	pp.load("z_length", extraction_params.z_length, L);
+        pp.load("num_points_phi", extraction_params.num_points_phi, 25);
+        pp.load("extraction_center", extraction_params.center,
+                {0.5 * L, 0.5 * L, 0.5 * L});
```
Same. {0.5*L} to 'center'

mirenradia replied:
I have removed the CylindricalExtractionTest as I think it's a bit redundant (see my comment above) but I have fixed this in the `SimulationParameters` for the SphericalExtractionTest

TaigoFr commented on Source/AMRInterpolator/SurfaceGeometry.hpp:
```diff
@@ -0,0 +1,64 @@
+{
+  public:
+    //! returns the grid spacing in u
+    virtual double du(int a_num_points_u) const = 0;
```
Most of the things in SphericalGeometry.hpp should have inline's, right?

mirenradia replied:
Unfortunately you can't inline pure virtual functions (as far as I'm aware). This goes to back to one of my questions in the original post about whether this generalisation using pointers to abstract base classes should just be changed to templating over a `SurfaceGeometry` class instead?

TaigoFr commented on Source/AMRInterpolator/SurfaceExtraction.hpp:
```diff
@@ -0,0 +1,201 @@
+  public:
+    //! Normal constructor which requires vars to be added after construction
+    //! using add_var or add_vars
+    SurfaceExtraction(SurfaceGeometry *a_geom_ptr, params_t &a_params,
```
If I can be really picky about this, SurfaceGeometry* and params_t can be const&, right? If yes, it also applies to the params_t in the other classes

mirenradia replied:
I agree with this mostly and have changed it but I don't see the point in having a const reference pointer so I just made it a const pointer.

TaigoFr commented on Source/AMRInterpolator/SurfaceExtraction.hpp:
```diff
@@ -0,0 +1,201 @@
+    void add_vars(const std::vector<std::pair<int, Derivative>> &a_vars);
+
+    //! Alternative constructor with a predefined vector of variables
+    SurfaceExtraction(SurfaceGeometry *a_geom_ptr, params_t &a_params,
```
Feature request: can we have a 3rd constructor in SurfaceExtraction that receives not a 'const std::vector<std::pair<int, Derivative>> &a_vars', but a 'const std::vector<int> &a_vars' that assumes no derivatives? So that I could make SurfaceExtraction(..., params, {c_chi, c_phi}, ...) without worrying about derivatives if I want none

mirenradia replied:
Done, also added a corresponding `SurfaceExtraction::add_vars(const std::vector<int> &a_vars)` overload which is used by the new constructor.

KAClough commented:
@mirenradia This looks great but I'd like to review in more detail... I'm travelling next week so if it can wait until the week after to be merged that would be good.

TaigoFr commented:
Regarding Miren's initial questions:
 1) Having a pointer to a SurfaceGeometry has to usefulness that it forces users to making all the virtual methods of the abstract class, which in a way wouldn't even be necessary if we had the templates, right? On the other hand, for code consistency, it would make more sense to have a template, as we already use template everywhere for this type of thing
 2) I think it looks quite general. If there are any problems in the future, I think it might be of people like Pau wanting to use it for other # of dimensions, not someone using a 2D parametrization that doesn't fit what you coded
 3) I don't think we should remove it. As they are still separate. If AMRInterpolator stops working, you want to know it's because of AMRInterpolator and not because of SurfaceExtraction. This is what I'd say
 4) I don't mind this at all. What type of user would care if the number of points is 50 or 51? You have a warning there, so it's fine by me

mirenradia commented:
> 1. Having a pointer to a SurfaceGeometry has to usefulness that it forces users to making all the virtual methods of the abstract class, which in a way wouldn't even be necessary if we had the templates, right? On the other hand, for code consistency, it would make more sense to have a template, as we already use template everywhere for this type of thing

If we switch to templating, I think we'll also be able to inline many of the `SurfaceGeometry` member functions.

KAClough commented:
Sorry for taking ages to review this. It is of course, very nice code :-) I especially appreciate how short WeylExtraction is now. This is going to be a much easier class for people to copy when they want to extract other things on a sphere, which is I suspect the main use case.

1. So, I think I agree with @TaigoFr - I would go for templating to be more consistent with the rest of the code. Would you then have WeylExtraction inheriting from SurfaceExtraction<SphericalGeometry>? For me this makes more sense, but then I am more used to templating than pointers.

2. It looks general to me and a great check is that for @amelialdrew it works for cylinders. I guess it has to be a differentiable surface… that is, the code won’t work for a box. But I think that is ok, as boxes are not very physical.

I would however agree with @amelialdrew that it would be more general if the params_t in SimulationParametersBase was SurfaceExtraction::params_t rather than specifying to spherical, so that they can be used as input for any Geometry class. You already inherit the SphericalExtraction::params_t from the SurfaceExtraction one, so this should work, although you might need to do something more to preserve backwards compatibility where people have num_points_phi etc in their params.txt.

3. I agree with @TaigoFr, we can never have too many tests :-) But I think we probably need a more rigorous one for the integration methods here. For example, instead of assigning a fixed value, it would be nice to assign something with a (theta, phi) dependence - e.g. a spherical harmonic, and then integrate it with the corresponding Y_lm using orthogonality to get an analytic result. Also, we should test all the methods, not just Simpson's one, and rather than setting a fixed bound on the error, we should show that it converges to zero at (roughly) the appropriate rate for each method when we increase the number of extraction points. This might also give us a good way of finding out roughly how many points is "good enough" for each method.

4. I agree that increasing the number of theta points by one (with user warning) is fine. As we discussed already, the worse issue with this integration is that by taking points at theta = 0 and pi, all the phi points here collapse onto the same physical point, meaning that if you choose num_points_phi = 50 you have 49 + 49 redundant points. As you say, they also aren’t used in principle for the integration, so it seems like there should be a solution for this where your endpoints are dtheta and (pi - dtheta)… I know it is not a big deal, but since it annoys me I will try to think of a non ugly fix.

A couple of other minor points:
5. cartesian_coord() -> get_chombo_grid_coords()  or get_interpolator_coords()? This returns the coords for the interpolator to use, whereas for me the cartesian ones should be those with reference to the centre (i.e., the transformed x,y,z of r, theta, phi). Otherwise it makes the code for the Y_lms a bit confusing (although the comments helped :-)). Maybe we can have two functions - one to get the cartesian coords (wrt centre), and one to get the grid coords? 

Also, sidenote, to verb or not to verb for these functions which just return a value?

6. Changing to m_p.extraction_params.center from m_p.extraction_params.extraction_center will potentially break user code (e.g. where they use Weyl4, and in some TaggingCriterion). I don’t really mind because I agree this is cleaner… but we should think if there is an easy way to prevent this.

7. I don’t really love the name “surface_param_value”… but I cannot think of anything better either, and it made sense to me, so I guess it is ok… I thought about “level_set_value” and "level_set_param" but that may be confusing too?

8. write_integral() -> write_single_integral() ? I’m just afraid of having the other function write_integrals() differing by only an “s".

mirenradia commented:
> Sorry for taking ages to review this. It is of course, very nice code :-) I especially appreciate how short WeylExtraction is now. This is going to be a much easier class for people to copy when they want to extract other things on a sphere, which is I suspect the main use case.

Thanks!

> 1. So, I think I agree with @TaigoFr - I would go for templating to be more consistent with the rest of the code. Would you then have WeylExtraction inheriting from `SurfaceExtraction<SphericalGeometry>`? For me this makes more sense, but then I am more used to templating than pointers.

Ok, I have switched to templating. This has also allowed me to inline all the `SurfaceGeometry` member functions which @TaigoFr and I wanted to do before.

> 2. It looks general to me and a great check is that for @amelialdrew it works for cylinders. I guess it has to be a differentiable surface… that is, the code won’t work for a box. But I think that is ok, as boxes are not very physical.

I think you technically should be able to implement extraction over a box with this class but you'd need to be very careful when writing the `SurfaceGeometry` class.
 
> I would however agree with @amelialdrew that it would be more general if the params_t in SimulationParametersBase was SurfaceExtraction::params_t rather than specifying to spherical, so that they can be used as input for any Geometry class. You already inherit the SphericalExtraction::params_t from the SurfaceExtraction one, so this should work, although you might need to do something more to preserve backwards compatibility where people have num_points_phi etc in their params.txt.

I slightly disagree with this but instead think `SimulationParametersBase` should be renamed instead (see my comment above).

> 1. I agree with @TaigoFr, we can never have too many tests :-) But I think we probably need a more rigorous one for the integration methods here. For example, instead of assigning a fixed value, it would be nice to assign something with a (theta, phi) dependence - e.g. a spherical harmonic, and then integrate it with the corresponding Y_lm using orthogonality to get an analytic result. Also, we should test all the methods, not just Simpson's one, and rather than setting a fixed bound on the error, we should show that it converges to zero at (roughly) the appropriate rate for each method when we increase the number of extraction points. This might also give us a good way of finding out roughly how many points is "good enough" for each method.

This is a good idea but will require more time to code so haven't done this yet.

> 2. I agree that increasing the number of theta points by one (with user warning) is fine. As we discussed already, the worse issue with this integration is that by taking points at theta = 0 and pi, all the phi points here collapse onto the same physical point, meaning that if you choose num_points_phi = 50 you have 49 + 49 redundant points. As you say, they also aren’t used in principle for the integration, so it seems like there should be a solution for this where your endpoints are dtheta and (pi - dtheta)… I know it is not a big deal, but since it annoys me I will try to think of a non ugly fix.

Uli suggested that, if the parity of the number of points is wrong, you can switch to Simpson's 3/8 rule (which has period 3 subintervals) for the first couple of subintervals and then use the normal Simpson's rule for the rest. The problem is that this would be a bit messy to implement with the `IntegrationMethod` class I have written. I think I'd have to virtualize some of the functions and write a child class in order to do this. Does anyone think this is worth it or shall we leave it as is?

Feel free to think of a solution to the redundant points in the spherical case. I couldn't think of one without breaking generality.

> 1. cartesian_coord() -> get_chombo_grid_coords()  or get_interpolator_coords()? This returns the coords for the interpolator to use, whereas for me the cartesian ones should be those with reference to the centre (i.e., the transformed x,y,z of r, theta, phi). Otherwise it makes the code for the Y_lms a bit confusing (although the comments helped :-)). Maybe we can have two functions - one to get the cartesian coords (wrt centre), and one to get the grid coords?

Yeah, this makes sense. I have changed it to `get_grid_coords()`.
 
> Also, sidenote, to verb or not to verb for these functions which just return a value?

This would be consistent with our coding conventions but since their implementation is so simple, I'm tempted to leave them as is. If you feel strongly about this, I won't object though.

> 1. Changing to m_p.extraction_params.center from m_p.extraction_params.extraction_center will potentially break user code (e.g. where they use Weyl4, and in some TaggingCriterion). I don’t really mind because I agree this is cleaner… but we should think if there is an easy way to prevent this.

I cheated and created a reference parameter which has the old name and just references the new parameter.

> 2. I don’t really love the name “surface_param_value”… but I cannot think of anything better either, and it made sense to me, so I guess it is ok… I thought about “level_set_value” and "level_set_param" but that may be confusing too?

I don't like this either but couldn't think of a better name. I think your suggestion `level_set_value` might be marginally better but it's still not the clearest so I have left it as is for the moment.

> 3. write_integral() -> write_single_integral() ? I’m just afraid of having the other function write_integrals() differing by only an “s".

Does this matter though? They both do the same thing and if you accidentally use the wrong one, it'll probably do what you want anyway. Since I've been looking at the code in the `Cell<data_t>` class recently, I noticed that one of the overloads is still `load_vars` even if you're loading just 1 variable. For consistency, perhaps I should even rename `write_integral` to make it an overload for `write_integrals`!

KAClough commented:
- The Simpson's 3/8 rule is a nice idea but I think we should leave it as it will make the code ugly. I think changing the number of points with a warning is totally fine. It will also make my quest to fix the degenerate end points harder :-)
-  I agree about the naming of SimulationParametersBase - I think we said it should be called GRParametersBase to match to GRAMRLevel naming... although I now think it should be divided into GRParametersBase and an (inherited) NRParametersBase as some of it is not used for fixed backgrounds... sadly it breaks the code to change it now... but with this classification, I think that it still should not specify spherical extraction as that is only one possible usage case in GR - e.g. for strings we use cylinders. I think that the conversion to an example specific surface should happen in the Example's SimulationParameters class.
- I don't feel strongly about the other naming issues, so I'm happy for you to leave those as they are.

Then, I think that all that is left to do is to make the tests, and for me to fix the end points thing (if I can find a nice solution)?

mirenradia commented:
> * I agree about the naming of SimulationParametersBase - I think we said it should be called GRParametersBase to match to GRAMRLevel naming... although I now think it should be divided into GRParametersBase and an (inherited) NRParametersBase as some of it is not used for fixed backgrounds... sadly it breaks the code to change it now... but with this classification, I think that it still should not specify spherical extraction as that is only one possible usage case in GR - e.g. for strings we use cylinders. I think that the conversion to an example specific surface should happen in the Example's SimulationParameters class.

The change to a general `SurfaceExtraction::params_t` rather than the spherical one in `SimulationParametersBase` would also break code though?
> Then, I think that all that is left to do is to make the tests, and for me to fix the end points thing (if I can find a nice solution)?

Yes, I agree with this. Writing the test is on my to do list so I should get round to it at some point.

KAClough commented:
> The change to a general SurfaceExtraction::params_t rather than the spherical one in SimulationParametersBase would also break code though?

I think even the change to the spherical one will break the code of people using that struct - as they will need to declare it as a SurfaceExtraction::params_t rather than an extraction_params_t when it is passed (as in Tagging criteria, for example).

mirenradia commented:
> I think even the change to the spherical one will break the code of people using that struct - as they will need to declare it as a SurfaceExtraction::params_t rather than an extraction_params_t when it is passed (as in Tagging criteria, for example).

Ah yes, that's true. I forgot about this!

mirenradia commented:
Ok, I've changed the test so it now checks the order of convergence of the Trapezium rule, Simpson's rule and Boole's rule. This is now ready for review :smile:

TaigoFr commented on Source/AMRInterpolator/IntegrationMethod.hpp:
```diff
@@ -0,0 +1,68 @@
+                         bool a_is_periodic) const
+    {
+        const int weight_index = a_index % m_num_weights;
+        const bool endpoint =
+            (a_index == 0 || a_index == a_num_points - 1) && !a_is_periodic;
+        if (m_is_closed && !endpoint && weight_index == 0)
+            return 2.0 * m_weights[weight_index];
+        else
+            return m_weights[weight_index];
```
I would appreciate a small comment added on this part.

mirenradia replied:
Fair enough. Will do.

TaigoFr commented on Source/AMRInterpolator/SphericalExtraction.hpp:
```diff
@@ -0,0 +1,139 @@
+        std::pair<std::vector<double>, std::vector<double>> integrals;
+        add_mode_integrand(es, el, em, a_function, integrals, a_method_theta,
+                           a_method_phi);
+        integrate();
```
I may be missing something, but if I want to integrate several modes and, instead of adding them with `add_mode_integrand` and then integrate, I just call this `integrate_mode` several times, this line here with `integrate` will cause it to integrate all the previous additions again, right? If I'm right, an easy solution is to use the `integrate` overload from `SurfaceExtraction` that receives the integrand directly (as that one clears previous integrands)

mirenradia replied:
Ah yes, I hadn't spotted that. Unfortunately your solution involves more code duplication so instead I will `clear` the relevant `std::vector`s at the start of this function.

TaigoFr commented on Tests/SphericalExtractionTest/SphericalExtractionTest.inputs:
```diff
@@ -0,0 +1,33 @@
+
+num_ghosts = 3
+max_level = 1
+regrid_interval = 1 1 1 1 0 0 0 0 0
+ref_ratio = 2 2 2 2 2 2 2 2 2
```
I guess you don't need this many `regrid_interval`s and `ref_ratios`s

mirenradia replied:
I guess not but it doesn't hurt in case someone wants to increase the number of levels in this test for whatever reason.

TaigoFr commented on Source/AMRInterpolator/CylindricalGeometry.hpp:
```diff
@@ -0,0 +1,85 @@
+class CylindricalGeometry
+{
+  private:
+    std::array<double, CH_SPACEDIM> m_center;
+    double m_z_length;
```
Can/should be `const`?

mirenradia replied:
I agree. Will change.

TaigoFr commented on Source/AMRInterpolator/SphericalGeometry.hpp:
```diff
@@ -0,0 +1,86 @@
+//! u = theta, v = phi
+class SphericalGeometry
+{
+  private:
+    std::array<double, CH_SPACEDIM> m_center;
```
Can/should be `const`?

mirenradia replied:
Ditto

TaigoFr commented on Tests/SphericalExtractionTest/SimulationParameters.hpp:
```diff
@@ -0,0 +1,63 @@
+    void read_params(GRParmParse &pp)
+    {
+        // extraction params
+        dx.fill(coarsest_dx);
+        origin.fill(coarsest_dx / 2.0);
```
I think `dx` and `origin` should be in `ChomboParameters`! I think Katy and I have discussed this, as at some point I also wanted to inherit only from `ChomboParameters` and these were the only parameters I needed. Plus, it makes sense and doesn't break anything

mirenradia replied:
I think this is done in the apparent horizon finder PR (COSMOS-CTC-Cambridge/GRChombo#175) so I won't change this here to avoid dealing with conflicts later. All the more reason to get that finished asap :stuck_out_tongue:

TaigoFr commented on Tests/SphericalExtractionTest/SphericalExtractionTest.cpp:
```diff
@@ -0,0 +1,195 @@
+#include "SphericalExtraction.hpp"
+#include "SphericalExtractionTestLevel.hpp"
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
```
I guess you don't need this here

mirenradia replied:
It's possible OpenMP is used in some of the AMR stuff. Although I didn't have problems removing it on my laptop, I will keep it here in case other systems/compilers have issues.

TaigoFr commented on Examples/BinaryBH/BinaryBHLevel.cpp:
```diff
@@ -142,11 +142,11 @@ void BinaryBHLevel::specificPostTimeStep()
+        if (m_level == m_p.extraction_params.min_extraction_level())
         {
             CH_TIME("WeylExtraction");
             // Now refresh the interpolator and do the interpolation
```
Is it a good time to remove the `m_gr_amr.m_interpolator->refresh();` from the GRAMRLevel? We should move it to inside the extraction... It  would be a good opportunity now. The only problem is that old codes would do the refresh twice. But...

mirenradia replied:
No, I think this is a bad idea as the `refresh()` is quite expensive due to the exchange of ghost cells and if one is doing multiple separate extractions consecutively (e.g. Weyl and ADM mass), one only wants to do this once.

TaigoFr replied:
True. So is this something that should be removed from inside the AHFinder? As one may want to extract Weyl and find the AH, so it is also being done twice in that case

mirenradia replied:
Probably but to be honest the AHFinder itself is going to be the main bottleneck there anyway.

TaigoFr commented on Source/AMRInterpolator/SphericalExtraction.hpp:
```diff
@@ -0,0 +1,139 @@
+        add_vars(a_vars);
+    }
+
+    SphericalExtraction(const params_t &a_params,
+                        const std::vector<int> &a_vars, double a_dt,
+                        double a_time, bool a_first_step,
+                        double a_restart_time = 0.0)
+        : SphericalExtraction(a_params, a_dt, a_time, a_first_step,
+                              a_restart_time)
+    {
+        add_vars(a_vars);
+    }
```
It it possible to have a constructor like this without the `a_first_step` (set to `a_dt==a_time` as is done for Weyl)? Just because I was testing it out, building my own example to test your code, and I had to go to the Weyl Extraction constructor to remember how this was done...

mirenradia replied:
I'm not going to do this because there are already quite a few constructors in this class and I don't want to confuse it more and also if you are writing your own child class of this, then you will probably copy and modify `WeylExtraction` anyway.

TaigoFr commented on Source/AMRInterpolator/SurfaceExtraction.impl.hpp:
```diff
@@ -0,0 +1,411 @@
+    int a_var, std::vector<double> &out_integrals,
+    const IntegrationMethod &a_method_u, const IntegrationMethod &a_method_v)
+{
+    CH_assert(a_var > 0 && a_var < m_vars.size());
```
I think this assertion is wrong. It should be `a_var >= 0`. I was getting wrong assertions in the example I built and couldn't understand why :P

mirenradia replied:
Yes I agree (or perhaps I was implicitly banning the extraction of `chi` :laughing: ).

mirenradia replied:
Yes I agree.

TaigoFr commented on Source/AMRInterpolator/IntegrationMethod.hpp:
```diff
@@ -0,0 +1,68 @@
+    //! Constructor
+    IntegrationMethod(const std::vector<double> &a_weights,
+                      bool a_is_closed = true)
+        : m_weights(a_weights), m_num_weights(a_weights.size()),
+          m_is_closed(a_is_closed)
+    {
+    }
```
Maybe add an assertion checking that `a_weights.size()>0`

mirenradia replied:
Sure

TaigoFr commented on Source/TaggingCriteria/ChiExtractionTaggingCriterion.hpp:
```diff
@@ -42,7 +42,7 @@ class ChiExtractionTaggingCriterion
     // The constructor
     ChiExtractionTaggingCriterion(const double dx, const int a_level,
                                   const int a_max_level,
```
I think `max_level` is not being used in this criterion

mirenradia replied:
Yes, I think this was accidentally added in #75. I won't change it to maintain backwards compatibility.

TaigoFr commented on Source/AMRInterpolator/SurfaceExtraction.hpp:
```diff
@@ -0,0 +1,165 @@
+    //! correspond to the order in which the desired var was added to this
+    //! object with add_var
+    void add_var_integrand(
+        int a_var, std::vector<double> &out_integrals,
```
I think `a_var` should be be the index in `UserVariables` and not the index in the variables you added with `add_var`. This is easy as you only need to check the variable requested is inside the `m_vars` vector. This was a cause of error in the example I built. If I do `add_var(c_rho)`, I would probably want to follow with `add_var_integrand(c_rho, ...)`.
PLUS. And this is a less trivial re-write of the class: I think it would be a better idea if "labels" were passed each time an integrand is added, and saved inside the class. And not in the end a full vector for `write_integrals`. Why? Well, the biggest reason is so that when I do `àdd_var_integrand` I don't have to specify the label, because from `variable_names` you already have it. Like this you would specify a label for normal integrands as usual and no need to specify it for variable integrands.

mirenradia replied:
I don't think these changes are compatible with some `var`s being derivatives.

TaigoFr replied:
We could automate it by adding a label as "d" + name + "/dx" or something. The cost would be that the user loses liberty to give personalized names. I'm not sure if this is what you meant by compatible.

mirenradia replied:
OK, what I meant is that it's not compatible with your first suggestion. I don't like removing the user's choice for `add_var_integrand` but I think choosing the label at the `add_integrand` stage is not a bad idea.

TaigoFr replied:
Ok. If this ends up being smth to do, we could leave it as a choice for the user, but have a default that gets automatically filled for the case of the `add_var_integrand`

mirenradia replied:
Actually, having looked at the code again, I don't want to do this because the `write_integrals` function is separated from the integration and this allows the user to choose which integrals are written (they might just want to use some in the code without writing). I think it will be too difficult to store labels at the `add_integrand` stage and then match them with integrals passed to `write_integrals` so therefore I propose keeping things as they are.

KAClough replied:
This seems reasonable.

TaigoFr commented:
Besides the code comment I wanted to leave, I build a small test to see if everything was working. I calculated the ADM mass of a Kerr BH with M=1, no spin. N = 64, L=128, max_level = 2, extraction at r=30, 40, 50 at level 1. Then 3-point Richardson extrapolation to infinity. Even though the resolution is awfully bad (I wanted a quick run on my laptop), the results are impressive:
 - old extraction: 1.5e-4 relative error (==> M=1.00015)
 - new extraction + trapezium method: 3.3e-4 (Miren claims that "The prefactor on the leading error term is bigger for Trapezium", so all good)
 - new extraction + simpson method (in theta, still trapezium in phi): 6e-7
 - new extraction + boole method: 2e-8
I'm amazed at this precision. Even if I extract at level 0 I get similarly amazing results. 
I didn't test convergence order of the methods because the Test that Miren built already proves it's good on that side. So overall, I'm very satisfied.

One more note:
 - I'd be happy if there was a default `execute_query` as we have for `WeylExctraction`. For example, for the ADM mass example I built, I had to insert the following code in my `specificPostTimeStep`:
```
m_gr_amr.m_interpolator->refresh();
SphericalExtraction my_extraction(
    m_p.extraction_params, (const std::vector<int> &){c_Madm}, m_dt,
    m_time, (m_dt == m_time), m_restart_time);
my_extraction.extract(m_gr_amr.m_interpolator);
std::vector<double> out_integrals;
my_extraction.add_var_integrand(0, out_integrals,
                                IntegrationMethod::simpson);
my_extraction.integrate();
my_extraction.write_integral("Integral", out_integrals);
```
It would be good if I users could simply do something as
```
SphericalExtraction my_extraction(
    m_p.extraction_params, (const std::vector<int> &){c_Madm}, m_dt,
    m_time, m_restart_time);
std::vector<double> out_integrals;
my_extraction.add_var_integrand(c_Madm, out_integrals,
                                IntegrationMethod::simpson);
my_extraction.execute_query(m_gr_amr.m_interpolator);
```
or something like that.

mirenradia commented on Source/AMRInterpolator/IntegrationMethodSetup.hpp:
```diff
@@ -0,0 +1,20 @@
+/* GRChombo
+ * Copyright 2012 The GRChombo collaboration.
+ * Please refer to LICENSE in GRChombo's root directory.
+ */
+
+#ifndef INTEGRATIONMETHODSETUP_HPP_
+#define INTEGRATIONMETHODSETUP_HPP_
+
+#include "IntegrationMethod.hpp"
+
+// define the static IntegrationMethods here
+const IntegrationMethod IntegrationMethod::trapezium({0.5});
+const IntegrationMethod IntegrationMethod::midpoint({1.0}, false);
+const IntegrationMethod IntegrationMethod::simpson({0.3333333333333333,
+                                                    1.3333333333333333});
+const IntegrationMethod
+    IntegrationMethod::boole({0.3111111111111111, 1.4222222222222222,
+                              0.53333333333333, 1.4222222222222222});
+
```
I think I tried putting these in a cpp file (after first trying putting them in the `IntegrationMethod.hpp` header file) but it didn't work for some reason and then saw this is what is done for the `static const` `Derivative`s so I copied it. I can't imagine a situation where this particular file will ever needed to be included elsewhere.

mirenradia commented on Source/AMRInterpolator/IntegrationMethod.hpp:
```diff
@@ -0,0 +1,68 @@
+  public:
+    //! Constructor
+    IntegrationMethod(const std::vector<double> &a_weights,
+                      bool a_is_closed = true)
```
I see your point. My thinking was that closed Newton-Cotes formulae are generally the default and are what one first meets/thinks of when considering numerical integration (e.g. the trapezium rule, Simpson's rule) whereas open formulae are somewhat less common. Like you said, I don't think this code will really be looked at again anyway.

mirenradia commented on Source/AMRInterpolator/IntegrationMethod.hpp:
```diff
@@ -0,0 +1,68 @@
+    {
+        if (m_is_closed && !a_is_periodic)
+        {
+            return (a_num_points % m_num_weights == 1 % m_num_weights);
```
Yes I agree. I will change it to your first option.

mirenradia commented:
> One more note:
> 
> * I'd be happy if there was a default `execute_query` as we have for `WeylExctraction`. For example, for the ADM mass example I built, I had to insert the following code in my `specificPostTimeStep`:
> 
> ```
> m_gr_amr.m_interpolator->refresh();
> SphericalExtraction my_extraction(
>     m_p.extraction_params, (const std::vector<int> &){c_Madm}, m_dt,
>     m_time, (m_dt == m_time), m_restart_time);
> my_extraction.extract(m_gr_amr.m_interpolator);
> std::vector<double> out_integrals;
> my_extraction.add_var_integrand(0, out_integrals,
>                                 IntegrationMethod::simpson);
> my_extraction.integrate();
> my_extraction.write_integral("Integral", out_integrals);
> ```
> 
> It would be good if I users could simply do something as
> 
> ```
> SphericalExtraction my_extraction(
>     m_p.extraction_params, (const std::vector<int> &){c_Madm}, m_dt,
>     m_time, m_restart_time);
> std::vector<double> out_integrals;
> my_extraction.add_var_integrand(c_Madm, out_integrals,
>                                 IntegrationMethod::simpson);
> my_extraction.execute_query(m_gr_amr.m_interpolator);
> ```
> 
> or something like that.

I think this might be confusing for anything that doesn't involve trivial integration of a single variable and users might use it unintentionally. Besides, the difference is 2 lines (given my explanation about the `refresh`).

TaigoFr commented:
I also wanted to add that @mirenradia removed OMP from the integration. I tested this with Chombo Timers and indeed it makes no difference. The bottleneck is in the AMRInterpolator (which doesn't have OMP, I'm not sure why or if it could, but that's another topic)

mirenradia commented:
Thanks for testing that!

KAClough commented:
Ok, I think Miren has suffered enough reviewing from us in return for his heroic efforts to improve the code, so I'm going to merge this :-)

KAClough commented:
Thanks also to @TaigoFr for his forensic reviewing :-)
